hilti/toolchain/src/compiler/scope-builder.cc

```diff
@@ -120,23 +118,42 @@ struct Visitor : visitor::PostOrder {
 
     void operator()(statement::Declaration* n) final { n->parent()->getOrCreateScope()->insert(n->declaration()); }
 
-    void operator()(statement::For* n) final { n->getOrCreateScope()->insert(n->local()); }
+    void operator()(statement::For* n) final {
+        n->getOrCreateScope()->insert(n->local());
+
+        // Also add this to the body to avoid redefinitions
+        n->body()->getOrCreateScope()->insert(n->local());
+    }
 
     void operator()(statement::If* n) final {
-        if ( n->init() )
-            n->getOrCreateScope()->insert(n->init());
+        if ( auto* init = n->init() ) {
+            n->getOrCreateScope()->insert(init);
+
+            // Also add this to the true body to avoid redefinitions
+            n->true_()->getOrCreateScope()->insert(init);
```

Comment:
We need it for the `else` branch as well since the following is valid Spicy code (modelling semantics also in e.g., C):

```ruby
if (local a = 0) {
    # Empty.
} else {
    print a;
}
```

Let's also add a test for this.

Reply:
Dang, I should've tested that, I thought about it but thought it'd be weird if the scope extended inside the `else` - my bad!